File: CPU_ISA_java/src/Instruction.java
public class Instruction {
    private int step;
    private String opcode;
    private Register register;
    private int clockCycles;
    private String operand;
    private int value;

    public Instruction(int step, String opcode, Register register, int clockCycles) {
        this(step, opcode, register, clockCycles, "");
    }

    public Instruction(int step, String opcode, Register register, int clockCycles, String operand) {
        this.step = step;
        this.opcode = opcode;
        this.register = register;
        this.clockCycles = clockCycles;
        this.value = register.getValue();
        this.operand = operand;
    }

    public int getStep() {
        return step;
    }

    public void setStep(int step) {
        this.step = step;
    }

    public String getOpcode() {
        return opcode;
    }

    public void setOpcode(String opcode) {
        this.opcode = opcode;
    }

    public Register getRegister() {
        return register;
    }

    public void setRegister(Register register) {
        this.register = register;
    }

    public int getClockCycles() {
        return clockCycles;
    }

    public void setClockCycles(int clockCycles) {
        this.clockCycles = clockCycles;
    }

    public String getOperand() {
        return operand;
    }

    public void setOperand(String operand) {
        this.operand = operand;
    }

    public int getValue() {
        return value;
    }

    public void setValue(int value) {
        this.value = value;
    }

    public String fiveBitOpcode() {
        switch (opcode) {
            case "mov":
                return "00001";
            case "add":
                return "00010";
            case "mul":
                return "00100";
            case "sub":
                return "00011";
            case "div":
                return "00101";
            default:
                return "end 0 0";
        }
    }

    public String to16BitValue() {
        String binaryValue = Integer.toBinaryString(value);
        if (value >= 0) {
            return String.format("%16s", binaryValue).replace(' ', '0');
        } else {
            return binaryValue.substring(Math.max(0, binaryValue.length() - 16));
        }
    }

    @Override
    public String toString() {
        String operandValue = operand.equals("") ? String.valueOf(value) : operand;
        return String.format("[%d] %-5s%-4s %-5s%-22s%d",
                step, opcode, register.getAddress(), operand,
                to16BitValue(), clockCycles);
    }
}

File: python/instruction.py
class Instruction:
    def __init__(self, step, opcode, register, clock_cycles, operand=""):
        self.step = step
        self.opcode = opcode
        self.register = register
        self.clock_cycles = clock_cycles
        self.value = register.value
        self.operand = operand

    def five_bit_opcode(self):
        opcode_map = {
            'mov': '00001',
            'add': '00010',
            'mul': '00100',
            'sub': '00011',
            'div': '00101'
        }
        return opcode_map.get(self.opcode, 'end 0 0')

    def __str__(self):
        operand_value = str(self.value) if self.operand == "" else self.operand
        return f"[{self.step}] {self.opcode:<5}{self.register.address:<4} {self.operand:<5}{self.register.to_16bit_value():<22}{self.clock_cycles}" 